### PR TITLE
openpgp: take the most recent self-signature per UID

### PR DESCRIPTION
**Pull request by maxtaco**
- As per RFC4880 Section 5.2.3.3

### CODE REVIEW

File: openpgp/keys.go [hunk 1]
```diff
@@ -361,7 +361,14 @@ EachPacket:
...
+			// we should take the newer one.
+			if current != nil &&
+				(current.SelfSignature == nil || pkt.CreationTime.After(current.SelfSignature.CreationTime)) &&
+				(pkt.SigType == packet.SigTypePositiveCert || pkt.SigType == packet.SigTypeGenericCert) &&
+				pkt.IssuerKeyId != nil &&
+				*pkt.IssuerKeyId == e.PrimaryKey.KeyId {
+
 				if err = e.PrimaryKey.VerifyUserIdSignature(current.Name, e.PrimaryKey, pkt); err == nil {
 
 					// Only set this self-sig once. IT can't be overwritten.
```
Comment (oconnor663) [inline openpgp/keys.go]: This comment sounds like it's no longer correct?

File: openpgp/keys.go [hunk 2]
```diff
@@ -361,7 +361,14 @@ EachPacket:
...
 			// First handle the case of: a self-signature, and no previous self-signature
 			// verified for this UID.
-			if current != nil && current.SelfSignature == nil && (pkt.SigType == packet.SigTypePositiveCert || pkt.SigType == packet.SigTypeGenericCert) && pkt.IssuerKeyId != nil && *pkt.IssuerKeyId == e.PrimaryKey.KeyId {
+			// According to RFC8440, Section 5.2.3.3, if there are several self-signatures,
+			// we should take the newer one.
+			if current != nil &&
+				(current.SelfSignature == nil || pkt.CreationTime.After(current.SelfSignature.CreationTime)) &&
+				(pkt.SigType == packet.SigTypePositiveCert || pkt.SigType == packet.SigTypeGenericCert) &&
+				pkt.IssuerKeyId != nil &&
+				*pkt.IssuerKeyId == e.PrimaryKey.KeyId {
```
Comment (oconnor663) [inline openpgp/keys.go:369]: Should we separate this out into a separate function? I think it's important to highlight that

```
current.SelfSignature == nil || pkt.CreationTime.After(current.SelfSignature.CreationTime)
```

is the part that implements the comment above. But the monolithic expression makes it hard to do that. Having each of these be a line in a function would also make future diffs easier to look at.

**Comment by oconnor663** [comment]
LGTM, with the two questions above.

**Comment by maxtaco** [comment]
I decided not to restructure, just to fix the comments

**Comment by maxtaco** [comment]
I tried to find a gif about how comments are always wrong but failed.

Thanks for the review!

**Pull request merged**